File: lib/data/database/daos/stats_dao.dart
import 'package:drift/drift.dart';
import '../app_database.dart';
import '../tables.dart';

part 'stats_dao.g.dart';

@DriftAccessor(tables: [UserStats])
class StatsDao extends DatabaseAccessor<AppDatabase> with _$StatsDaoMixin {
  StatsDao(super.db);

  Future<UserStatEntity> getStats() async {
    final result = await (select(userStats)..where((t) => t.id.equals('default'))).getSingleOrNull();
    if (result == null) {
      await into(userStats).insert(UserStatsCompanion.insert(id: const Value('default')));
      return (await (select(userStats)..where((t) => t.id.equals('default'))).getSingle());
    }
    return result;
  }

  Stream<UserStatEntity> watchStats() {
    return (select(userStats)..where((t) => t.id.equals('default'))).watchSingleOrNull().map((s) => s!);
  }

  Future<void> updateStats({
    int? totalXp,
    int? currentLevel,
    int? streakDays,
    DateTime? lastActiveDate,
  }) async {
    await (update(userStats)..where((t) => t.id.equals('default'))).write(
      UserStatsCompanion(
        totalXp: totalXp != null ? Value(totalXp) : const Value.absent(),
        currentLevel: currentLevel != null ? Value(currentLevel) : const Value.absent(),
        streakDays: streakDays != null ? Value(streakDays) : const Value.absent(),
        lastActiveDate: lastActiveDate != null ? Value(lastActiveDate) : const Value.absent(),
        updatedAt: Value(DateTime.now()),
      ),
    );
  }
}

File: app/src/main/java/com/salesquest/sales_quest/data/dao/StatsDao.kt
package com.salesquest.sales_quest.data.dao

import androidx.room.Dao
import androidx.room.Insert
import androidx.room.OnConflictStrategy
import androidx.room.Query
import com.salesquest.sales_quest.data.entity.UserStatEntity
import kotlinx.coroutines.flow.Flow

@Dao
interface StatsDao {

    @Query("SELECT * FROM user_stats WHERE id = 'default'")
    suspend fun getStats(): UserStatEntity?

    @Query("SELECT * FROM user_stats WHERE id = 'default'")
    fun watchStats(): Flow<UserStatEntity?>

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun insertStats(stats: UserStatEntity)

    @Query(
        "UPDATE user_stats SET totalXp = :totalXp, currentLevel = :currentLevel, " +
            "streakDays = :streakDays, lastActiveDate = :lastActiveDate, updatedAt = :updatedAt WHERE id = 'default'"
    )
    suspend fun updateStats(
        totalXp: Int,
        currentLevel: Int,
        streakDays: Int,
        lastActiveDate: Long?,
        updatedAt: Long
    )
}
